x/wasm/relay_pingpong_test.go

```diff
@@ -0,0 +1,323 @@
+	}
+	// start game
+	var start startGame
+	if err := json.Unmarshal(data, &start); err != nil {
+		return nil, 0, err
+	}
+
+	ctx := p.chain.GetContext()
+	channelCap, ok := p.chain.App.WasmKeeper.ScopedKeeper.GetCapability(ctx, host.ChannelCapabilityPath(start.Source.Port, start.Source.Channel))
+	if !ok {
+		return nil, 0, sdkerrors.Wrap(channeltypes.ErrChannelCapabilityNotFound, "module does not own channel capability")
+	}
+	service := NewHit(p.actor, start.Value)
+	p.t.Logf("[%s] starting game with: %d: %v\n", p.actor, start.Value, service)
+
+	var seq uint64 = 1
+	packet := channeltypes.NewPacket(service.GetBytes(), seq, start.Source.Port, start.Source.Channel, start.CounterParty.Port, start.CounterParty.Channel, doNotTimeout, 0)
+	err := p.chain.App.WasmKeeper.ChannelKeeper.SendPacket(ctx, channelCap, packet)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	p.IncrementCounter(sentBallsCountKey, store)
+	store.Set(lastBallSentKey, sdk.Uint64ToBigEndian(start.Value))
+	return &cosmwasmv2.HandleResponse{}, 0, nil
+}
+
+func (p player) OnIBCChannelOpen(hash []byte, params cosmwasmv2.Env, channel cosmwasmv2.IBCChannel, store prefix.Store, api cosmwasm.GoAPI, querier wasmkeeper.QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasmv2.IBCChannelOpenResponse, uint64, error) {
+	if channel.Version != p.actor {
+		return &cosmwasmv2.IBCChannelOpenResponse{Success: false, Reason: fmt.Sprintf("expected %q but got %q", p.actor, channel.Version)}, 0, nil
+	}
+	return &cosmwasmv2.IBCChannelOpenResponse{Success: true}, 0, nil
+}
+
+func (p player) OnIBCChannelConnect(hash []byte, params cosmwasmv2.Env, channel cosmwasmv2.IBCChannel, store prefix.Store, api cosmwasm.GoAPI, querier wasmkeeper.QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasmv2.IBCChannelConnectResponse, uint64, error) {
+	return &cosmwasmv2.IBCChannelConnectResponse{}, 0, nil
+}
+
+func (p player) OnIBCChannelClose(ctx sdk.Context, hash []byte, params cosmwasmv2.Env, channel cosmwasmv2.IBCChannel, meter sdk.GasMeter, gas uint64) (*cosmwasmv2.IBCChannelCloseResponse, uint64, error) {
+	panic("implement me")
+}
+
+var ( // store keys
+	lastBallSentKey        = []byte("lastBallSent")
+	lastBallReceivedKey    = []byte("lastBallReceived")
+	sentBallsCountKey      = []byte("sentBalls")
+	receivedBallsCountKey  = []byte("recvBalls")
+	confirmedBallsCountKey = []byte("confBalls")
+)
+
+func (p player) OnIBCPacketReceive(hash []byte, params cosmwasmv2.Env, packet cosmwasmv2.IBCPacket, store prefix.Store, api cosmwasm.GoAPI, querier wasmkeeper.QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasmv2.IBCPacketReceiveResponse, uint64, error) {
+	// parse received data and store
+	var receivedBall hit
+	if err := json.Unmarshal(packet.Data, &receivedBall); err != nil {
+		return &cosmwasmv2.IBCPacketReceiveResponse{
+			Acknowledgement: hitAcknowledgement{Error: err.Error()}.GetBytes(),
+			// no hit msg, we stop the game
+		}, 0, nil
+	}
+
+	otherCount := receivedBall[counterParty(p.actor)]
```

Comment:
Let's add a max hits (3 or 4) and if this is higher, then we return `IBCPacketReceiveResponse` with a "failed" response. In this case, we do not initiate another round. It will allow us to test error handling in acknowledge.